### PR TITLE
bazel: include zoekt version stamp in github.com/sourcegraph/zoekt.Version

### PR DESCRIPTION
**Pull request by Strum355**
As we were including the `go_binary`'s directly in the images, we couldn't set the x_defs attr in order to stamp. Given this, we patch each of the `go_binary` rules to include the x_defs.

## Test plan

```sh
$ VERSION=6.9.50 bazel build @com_github_sourcegraph_zoekt//cmd/zoekt-webserver --stamp --workspace_status_command=./dev/bazel_stamp_vars.sh
...
$ strings bazel-bin/external/com_github_sourcegraph_zoekt/cmd/zoekt-webserver/zoekt-webserver_/zoekt-webserver | grep 8cf8
v0.0.0-20240327102325-8cf8887a903a
```

### CODE REVIEW

**Review by burmudar — APPROVED** [review]

**Review by keegancsmith — COMMENTED** [review]
This is rockstar stuff, and I find it highly satisfying seeing how this all fits together and I really want the version set again...

However, this seems like a lot of complexity for what we gain:
- patch files
- repetitive x_defs
- running go list inside of our pipeline generation
- needing to understand all these moving parts in the future

Personally unless there was a simpler way to achieve this I'm not sure this is worth the cost?

**Comment by Strum355** [comment]
> This is rockstar stuff, and I find it highly satisfying seeing how this all fits together and I really want the version set again...
> 
> However, this seems like a lot of complexity for what we gain:
> 
> * patch files
> * repetitive x_defs
> * running go list inside of our pipeline generation
> * needing to understand all these moving parts in the future
> 
> Personally unless there was a simpler way to achieve this I'm not sure this is worth the cost?

I think at a minimum patches (potentially adding x_defs directly in the patches instead) and `go list` may be a minimum that we need here, would that be a more reasonable cost? Factoring out `go list` may be a larger time investment than the benefit

**Comment by Strum355** [comment]
Just to be clear on the above, I'm proposing making more zoekt-specific changes rather than a more general solution, which would mean we could reduce the diff to 1) patches to add x_defs directly 2) go list, reducing/eliminating the change to macro.bzl and all the `DEPS` variables

**Comment by keegancsmith** [comment]
> `go list`

Is there no way we can get this information from bazel? It is specified in `deps.bzl`. Is it maybe worth specifying the zoekt dep outside of the normal deps stuff to get what we want?

TBH if go list is super fast on CI, then I suppose this is fine. Otherwise I would just parse deps.bzl or go.mod for a faster response. It wouldn't be the first time I ran awk on our go.mod :)

> patches

The patches look a little gnarly, but acceptable if easy to regenerate.


---

Here is another idea we could try. We can make the main packages super lightweight in Zoekt, then actually have lightweight wrappers inside of the sourcegraph repo. So instead of building the main package directly from zoekt, it's inside of our monorepo => go_rules doesn't make our lives terrible for x_defs.

But yeah, I'm also totally happy with adding a file or environment variable to docker images with version info to avoid bad cache invalidation. Practically speaking we don't need to stamp the binaries, but just want to report something reasonable in our usual runtime envs.

But overall the version is not super critical information to have inside of zoekt. Which is why I am hesitant about introducing complexity that is spread out.

**Comment by Strum355** [comment]
> Is there no way we can get this information from bazel? It is specified in `deps.bzl`. Is it maybe worth specifying the zoekt dep outside of the normal deps stuff to get what we want?
> 
> TBH if go list is super fast on CI, then I suppose this is fine. Otherwise I would just parse deps.bzl or go.mod for a faster response. It wouldn't be the first time I ran awk on our go.mod :)

Invoking the pipeline generator takes ~1s, this includes running `go mod list`, so Im fine with this in CI 

> The patches look a little gnarly, but acceptable if easy to regenerate.

`echo $(bazel info output_base)/external/com_github_sourcegraph_zoekt)` will get you the directory from which you can work from, you can copy this & `git init && git add --all && git commit --allow-empty-message` to have a slate to start from. Then its the typical `git diff`

> Here is another idea we could try. We can make the main packages super lightweight in Zoekt, then actually have lightweight wrappers inside of the sourcegraph repo. So instead of building the main package directly from zoekt, it's inside of our monorepo => go_rules doesn't make our lives terrible for x_defs.

You could definitely do that, I feel like itd be more work than the current diff if you want to take a look at the alternate approach Ive pushed

> But yeah, I'm also totally happy with adding a file or environment variable to docker images with version info to avoid bad cache invalidation. Practically speaking we don't need to stamp the binaries, but just want to report something reasonable in our usual runtime envs.

Bazel has a concept of "stable" vs "volatile" stamp vars, where the former don't invalidate cache if they change (so only changes to the actual closure will cause a rebuild). Thats achieved by prefixing with `STABLE_`, which Ive done here

**Comment by Strum355** [comment]
Mixed up stable vs volatile :clueless: lemme address that real quick

**Comment by keegancsmith** [comment]
Thanks for updating and working further on this. I won't have time today to look deeply. Feel free to merge if other dev-infra people are happy. But I'll be back after easter weekend and will take a look :)

**Comment by jhchabran** [comment]
I'm a bit ambivalent on the solution proposed here. On one hand, it works, and solves the now problem, but on the other hand, while this is leveraging what we can do with Bazel, it's still complex, especially since the CI pipeline now has the responsibility of coming up with that version number, which is the biggest smell conceptually imo. 

The problem is that we're doing two things here, first we're building the zoekt binaries and second we're also using the packages from that go module. If we had the first one only, I'd have been keen to set up a separate pipeline and fetch binaries from there. 

Perhaps, still a workaround, but less debt incurring is that we could remove the part in the pipeline generator by making the stamping script able to find the zoekt version itself: 

```
echo ZOEKT_VERSION $(go list -m -f '{{.Version}}' github.com/sourcegraph/zoekt)  
```

And it's pretty fast: 

```sh
~/work/sourcegraph  main $ time go list -m -f '{{.Version}}' github.com/sourcegraph/zoekt
v0.0.0-20240327102325-8cf8887a903a
go list -m -f '{{.Version}}' github.com/sourcegraph/zoekt  0.01s user 0.02s system 20% cpu 0.167 total
```

And if somehow running `go` directly is a problem, we could always fall back to grepping the `go.mod`, which while not very nice, would work. Still, we would remove the pipeline generator from the equation, which is more robust on the long term as it's not something we want to keep growing, but the opposite.

**Comment by jhchabran** [comment]
> But overall the version is not super critical information to have inside of zoekt. Which is why I am hesitant about introducing complexity that is spread out.

I agree with this, we're paying quite the price by having to stamp binaries at the end, which while fine on zoekt as it's not going to move too often, is still problematic on sourcegraph/sourcegraph as it prevents us to cache a lot of outputs when building the stamped OCI images. 

I'd be more keen on injecting a var in the containers, which would ultimately still be done at build time but would be a much faster operation. 

Anyway, that's another topic :)

**Comment by Strum355** [comment]
> we're paying quite the price by having to stamp binaries at the end, which while fine on zoekt as it's not going to move too often, is still problematic on sourcegraph/sourcegraph as it prevents us to cache a lot of outputs when building the stamped OCI images.

I dont understand this. The stamp value only changes when we bump zoekt, at which stage the OCI image is going to have to be rebuilt anyways because, irregardless of the stamp, the zoekt binary has changed anyway

> And if somehow running go directly is a problem, we could always fall back to grepping the go.mod, which while not very nice, would work.

Running go is the issue because 1) the agents don't have go and 2) you cant `bazel run` during a `bazel build`. The grep solution seems most simple and reliable if we're ruling out `go list`, I'm fine with using grep

**Comment by Strum355** [comment]
> > I dont understand this. The stamp value only changes when we bump zoekt, at which stage the OCI image is going to have to be rebuilt anyways because, irregardless of the stamp, the zoekt binary has changed anyway
> 
> I was talking about sg/sg here, it's unreleated to zoekt, I was making a larger observation on that approach

"Volatile" stamp values (as defined by bazel, as opposed to "stable" stamp values) don't invalidate cache by design, so I'm unsure what approach youre referencing here

**Comment by jhchabran** [comment]
@Strum355 I missed the fact that we're using _volatile_ here for Zoekt. But regardless, my point was about sourcegraph/sourcegraph and stamping the actual version number in the binaries (gitserver, frontend, etc...), not zoekt. Those version numbers are not _volatile_, they're stable (talking about [this](https://github.com/sourcegraph/sourcegraph/pull/61464/files#diff-9eda1ac3d8b1e55fb0f146381c826f7108cd5e61c59cc39d5730abcb10b87782R3)) 

@keegancsmith for now, I think if handle the stamping for Zoekt as it's done in this PR, with getting the zoekt version from the status script and not the pipeline generator, we're good to go for the time being. 

I think that finding a long term solution to this particular pattern should be pursued in separately. I agree that it's not complicated to make zoekt read a file in the container to get its version, but implementing that a few days before the next release doesn't leave us enough room to focus on shipping a solution that is well thought out.

**Comment by jhchabran** [comment]
> Is it too gruesome to have a rule which just greps or runs go list on our go.mod file?

The problem is that `x_defs` cannot be set with anything else than stamping, [we can't use](https://github.com/bazelbuild/rules_go/blob/master/go/private/rules/binary.bzl#L261-L265) a _target_ to feed that value there, because it's a `string_dict` and not a `label`.

**Comment by keegancsmith** [comment]
I think I'm gonna bow out on this review now since its upto the bazel experts to make the decisions :) Either way thanks so much for the work put in and whatever we end up with. Just ping me if needed.

**Comment by Strum355** [comment]
@jhchabran i've removed the pipeline generator stage, using `grep` + `awk` instead in the stampvars script. LMKWYT

**Review by keegancsmith — COMMENTED** [review]
I don't know about you, but this is a nice tasty treat of a minimal change. Will leave approval to the bazel mages.

**Review by jhchabran — APPROVED** [review]
Approving, just in case, don't bother going again through the approval cycle if add the test that Keegan mentioned 😊

File: dev/bazel_stamp_vars.sh [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 echo STABLE_VERSION "$VERSION"
 echo VERSION_TIMESTAMP "$(date +%s)"
+echo ZOEKT_VERSION "$(grep sourcegraph/zoekt go.mod | awk '{ print $2 }')"
```
[the remark this answers is not in the document]
Comment (Strum355) [reply dev/bazel_stamp_vars.sh:5]: Included a test here, I think this should cover it https://github.com/sourcegraph/sourcegraph/pull/61464/commits/099c620638d809eb0f102865a66817296173427f
cc @jhchabran

[the remark this answers is not in the document]
Comment (jhchabran) [reply dev/bazel_stamp_vars.sh:5]: Neat

[the remark this answers is not in the document]
Comment (Strum355) [reply dev/bazel_stamp_vars.sh:5]: Lol it just caught that I pushed with an (intentional for testing) bug :D

**Pull request merged**